File: src/ros2_bridge/mappers/visualization_msgs/interactive_marker_init.rs
//! Typed mapper for `visualization_msgs/msg/InteractiveMarkerInit`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn interactive_marker_init_to_bus(msg: ros_env::visualization_msgs::msg::InteractiveMarkerInit) -> crate::visualization_msgs::msg::v1::InteractiveMarkerInit {
    crate::visualization_msgs::msg::v1::InteractiveMarkerInit {
        server_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.server_id),
        seq_num: msg.seq_num,
        markers: msg.markers.into_iter().map(crate::ros2_bridge::mappers::visualization_msgs::interactive_marker::interactive_marker_to_bus).collect(),
    }
}

pub(crate) fn interactive_marker_init_to_ros(bus: crate::visualization_msgs::msg::v1::InteractiveMarkerInit) -> ros_env::visualization_msgs::msg::InteractiveMarkerInit {
    ros_env::visualization_msgs::msg::InteractiveMarkerInit {
        server_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.server_id),
        seq_num: bus.seq_num,
        markers: bus.markers.into_iter().map(crate::ros2_bridge::mappers::visualization_msgs::interactive_marker::interactive_marker_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct VisualizationMsgsInteractiveMarkerInitMapper;

impl TypedTopicMapper for VisualizationMsgsInteractiveMarkerInitMapper {
    type Ros = ros_env::visualization_msgs::msg::InteractiveMarkerInit;
    type Bus = crate::visualization_msgs::msg::v1::InteractiveMarkerInit;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(interactive_marker_init_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(interactive_marker_init_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/visualization_msgs/interactive_marker_init.py
"""Generated mapper for `visualization_msgs/msg/InteractiveMarkerInit`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.visualization_msgs.interactive_marker import interactive_marker_to_bus, interactive_marker_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.visualization_msgs.msg.v1 import InteractiveMarkerInit as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def interactive_marker_init_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.server_id = str(msg.server_id)
    bus.seq_num = msg.seq_num
    bus.markers.extend([interactive_marker_to_bus(x) for x in msg.markers])
    return bus


def interactive_marker_init_to_ros(bus):
    from visualization_msgs.msg import InteractiveMarkerInit as RosMsg

    out = RosMsg()
    out.server_id = str(bus.server_id)
    out.seq_num = bus.seq_num
    out.markers = [interactive_marker_to_ros(x) for x in bus.markers]
    return out


class VisualizationMsgsInteractiveMarkerInitMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from visualization_msgs.msg import InteractiveMarkerInit as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return interactive_marker_init_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return interactive_marker_init_to_ros(bus)
